gcc error "invalid use of void expression" from (mis)use of nil

### Example
```nim
when defined(direct):
  echo cast[ptr int](nil)[]

when defined(indirect):
  var x: ref int = nil
  echo cast[ptr int](x)[]
```

### Current Output -d:indirect
```
Traceback (most recent call last)
/home/jfondren/nim/learn/error1.nim(6) error1
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

### Current Output -d:direct
```
/home/jfondren/.cache/nim/error1_d/@merror1.nim.c: In function ‘NimMainModule’:
/home/jfondren/.cache/nim/error1_d/@merror1.nim.c:152:24: error: invalid use of void expression
  152 |  T1_[0] = nimIntToStr((*NIM_NIL));
      |                       ~^~~~~~~~~
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/home/jfondren/.choosenim/toolchains/nim-1.4.0/lib -I/home/jfondren/nim/learn -o /home/jfondren/.cache/nim/error1_d/@merror1.nim.c.o /home/jfondren/.cache/nim/error1_d/@merror1.nim.c' failed with exit code: 1
```

### Additional Information
This isn't a blocker to anything. This is my riffing off someone's `nil[]` joke from IRC.
```
$ nim -v
Nim Compiler Version 1.4.0 [Linux: amd64]
Compiled at 2020-10-16
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: bdcd87afca238a0a7b2c70971827cf9172817b12
active boot switches: -d:release
```

--- a/compiler/semfold.nim
+++ b/compiler/semfold.nim
@@ -674,7 +674,7 @@ proc getConstExpr(m: PSym, n: PNode; idgen: IdGenerator; g: ModuleGraph): PNode 
   of nkCast:
     var a = getConstExpr(m, n[1], idgen, g)
     if a == nil: return
-    if n.typ != nil and n.typ.kind in NilableTypes:
+    if n.typ != nil and n.typ.kind in NilableTypes and a.kind != nkNilLit:
       # we allow compile-time 'cast' for pointer types:
       result = a
       result.typ = n.typ

--- a/lib/posix/posix_linux_amd64_consts.nim
+++ b/lib/posix/posix_linux_amd64_consts.nim
@@ -365,7 +365,7 @@ const SCHED_RR* = cint(2)
 const SCHED_OTHER* = cint(0)
 
 # <semaphore.h>
-const SEM_FAILED* = cast[pointer]((nil))
+const SEM_FAILED* = nil
 
 # <signal.h>
 const SIGEV_NONE* = cint(1)

--- a/tests/ccgbugs/t15623.nim
+++ b/tests/ccgbugs/t15623.nim
@@ -1,0 +1,10 @@
+discard """
+  action: "compile"
+"""
+
+block:
+  echo cast[ptr int](nil)[]
+
+block:
+  var x: ref int = nil
+  echo cast[ptr int](x)[]

--- a/tests/ccgbugs/t15623.nim
+++ b/tests/ccgbugs/t15623.nim
@@ -2,6 +2,7 @@ discard """
   action: "compile"
 """
 
+# bug #15623
 block:
   echo cast[ptr int](nil)[]
 

--- a/tests/ccgbugs/t15623_2.nim
+++ b/tests/ccgbugs/t15623_2.nim
@@ -1,0 +1,13 @@
+discard """
+  output: '''0
+0
+'''
+"""
+
+# bug #15623
+block:
+  echo cast[int](cast[ptr int](nil))
+
+block:
+  var x: ref int = nil
+  echo cast[int](cast[ptr int](x))

